controlplane/api/v1beta1/conversion.go

```diff
@@ -0,0 +1,89 @@
+	out.Replicas = in.Replicas
+	out.Version = in.Version
+	if err := Convert_v1beta1_KThreesConfigSpec_To_v1beta2_KThreesConfigSpec(&in.KThreesConfigSpec, &out.KThreesConfigSpec, s); err != nil {
+		return fmt.Errorf("converting KThreesConfigSpec field from v1beta1 to v1beta2: %w", err)
+	}
+	out.UpgradeAfter = in.UpgradeAfter
+	if err := Convert_v1beta1_KThreesControlPlaneMachineTemplate_To_v1beta2_KThreesControlPlaneMachineTemplate(&in.MachineTemplate, &out.MachineTemplate, s); err != nil {
+		return fmt.Errorf("converting KThreesControlPlaneMachineTemplate field from v1beta1 to v1beta2: %w", err)
+	}
+	out.MachineTemplate.NodeDrainTimeout = in.NodeDrainTimeout
+	out.MachineTemplate.InfrastructureRef = in.InfrastructureTemplate
+	out.RemediationStrategy = (*cabp3v1.RemediationStrategy)(unsafe.Pointer(in.RemediationStrategy))
+	return nil
+}
+
+func Convert_v1beta2_KThreesControlPlaneSpec_To_v1beta1_KThreesControlPlaneSpec(in *cabp3v1.KThreesControlPlaneSpec, out *KThreesControlPlaneSpec, s conversion.Scope) error { //nolint: stylecheck
+	out.Replicas = in.Replicas
+	out.Version = in.Version
+	out.InfrastructureTemplate = in.MachineTemplate.InfrastructureRef
+	if err := Convert_v1beta2_KThreesConfigSpec_To_v1beta1_KThreesConfigSpec(&in.KThreesConfigSpec, &out.KThreesConfigSpec, s); err != nil {
+		return fmt.Errorf("converting KThreesConfigSpec field from v1beta2 to v1beta1: %w", err)
+	}
+	out.UpgradeAfter = in.UpgradeAfter
+	out.NodeDrainTimeout = in.MachineTemplate.NodeDrainTimeout
+	if err := Convert_v1beta2_KThreesControlPlaneMachineTemplate_To_v1beta1_KThreesControlPlaneMachineTemplate(&in.MachineTemplate, &out.MachineTemplate, s); err != nil {
+		return fmt.Errorf("converting KThreesControlPlaneMachineTemplate field from v1beta2 to v1beta1: %w", err)
+	}
+	out.RemediationStrategy = (*RemediationStrategy)(unsafe.Pointer(in.RemediationStrategy))
+	return nil
+}
+
+func Convert_v1beta2_KThreesControlPlaneMachineTemplate_To_v1beta1_KThreesControlPlaneMachineTemplate(in *cabp3v1.KThreesControlPlaneMachineTemplate, out *KThreesControlPlaneMachineTemplate, s conversion.Scope) error { //nolint: stylecheck
+	out.ObjectMeta = in.ObjectMeta
+	return nil
+}
+
+func Convert_v1beta1_KThreesConfigSpec_To_v1beta2_KThreesConfigSpec(in *bootstrapv1beta1.KThreesConfigSpec, out *bootstrapv1beta2.KThreesConfigSpec, s conversion.Scope) error { //nolint: stylecheck
+	return bootstrapv1beta1.Convert_v1beta1_KThreesConfigSpec_To_v1beta2_KThreesConfigSpec(in, out, s)
+}
+
+func Convert_v1beta2_KThreesConfigSpec_To_v1beta1_KThreesConfigSpec(in *bootstrapv1beta2.KThreesConfigSpec, out *bootstrapv1beta1.KThreesConfigSpec, s conversion.Scope) error { //nolint: stylecheck
+	return bootstrapv1beta1.Convert_v1beta2_KThreesConfigSpec_To_v1beta1_KThreesConfigSpec(in, out, s)
+}
+
+// ConvertTo converts the v1beta1 KThreesConfig receiver to a v1beta2 KThreesConfig.
+func (in *KThreesControlPlane) ConvertTo(dstRaw ctrlconversion.Hub) error {
```

Comment:
You'll need to restore destination version from the annotation. This means you will need this:

```go
        restored := &cabp3v1.KThreesControlPlane{}
	if ok, err := utilconversion.UnmarshalData(r, restored); err != nil || !ok {
		return err
	}
```

And then you can use **restored** to set the values for any new fields on **dst**. See [this](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controlplane/eks/api/v1beta1/conversion.go#L29) as an example.

Reply:
This is now failing in tests, I'm not sure I need it. 
I think I need the MarshalData, but not sure the purpose of unmarshalling it since I can get all information from the previous api object.

Reply:
Sorry I didn't reload your last comment. 
Then yes I don't need it in this case. I'll remove it.